airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLBulkLoadHandler.kt

```diff
@@ -0,0 +1,244 @@
+        fullyQualifiedTableName: String,
+        dataFilePath: String,
+        formatFilePath: String,
+        rowsPerBatch: Long? = null
+    ): String {
+        val quotedTableName = quoteIdentifier(fullyQualifiedTableName)
+        // The ROWS_PER_BATCH hint can help optimize the bulk load.
+        // If not provided, it won't be included in the statement.
+        val rowBatchClause = rowsPerBatch?.let { "ROWS_PER_BATCH = $it," } ?: ""
+        return StringBuilder()
+            .apply {
+                append("BULK INSERT $quotedTableName\n")
+                append("FROM '$dataFilePath'\n")
+                append("WITH (\n")
+                append("\tCODEPAGE = '$CODE_PAGE',\n")
+                append("\tDATA_SOURCE = '$bulkUploadDataSource',\n")
+                append("\tFORMATFILE_DATA_SOURCE = '$bulkUploadDataSource',\n")
+                append("\tFIRSTROW = 2,\n")
+                append("\tFORMAT = '$FILE_FORMAT',\n")
+                append("\tFORMATFILE = '$formatFilePath',\n")
+                append("\t$rowBatchClause\n")
+                append("\tKEEPNULLS\n")
+                append(")")
+            }
+            .toString()
+            .trimIndent()
+    }
+
+    /** Creates a Global temp table by cloning the column structure from the main table. */
+    private fun createTempTable(conn: Connection, tempTableName: String) {
+        val createTempTableSql =
+            """
+            SELECT TOP 0 *
+            INTO $tempTableName
+            FROM $schemaName.$mainTableName
+        """.trimIndent()
+
+        conn.prepareStatement(createTempTableSql).use { stmt -> stmt.executeUpdate() }
+        conn.commit()
+    }
+
+    /**
+     * Builds a MERGE statement using the provided PK and non-PK columns, always quoting column
+     * names to avoid keyword conflicts.
+     */
+    private fun buildMergeSql(
+        tempTableName: String,
+        primaryKeyColumns: List<String>,
+        nonPkColumns: List<String>
+    ): String {
+        val quotedTableName = quoteIdentifier("$schemaName.$mainTableName")
+        // 1. ON condition:
+        //    e.g. Target.[Pk1] = Source.[Pk1] AND Target.[Pk2] = Source.[Pk2]
+        val onCondition = primaryKeyColumns.joinToString(" AND ") { "Target.[$it] = Source.[$it]" }
+
+        // 2. All columns for the INSERT statement (quoted)
+        val allColumns = primaryKeyColumns + nonPkColumns
+        val allColumnsCsv = allColumns.joinToString(", ") { "[$it]" }
+
+        // 3. The UPDATE assignments for non-PK columns
+        //    e.g. Target.[col1] = Source.[col1], Target.[col2] = ...
+        val updateAssignments = nonPkColumns.joinToString(", ") { "Target.[$it] = Source.[$it]" }
+
+        // 4. The VALUES for the INSERT statement
+        //    e.g. Source.[Pk1], Source.[col1], Source.[col2]...
+        val sourceColumnsCsv = allColumns.joinToString(", ") { "Source.[$it]" }
+
+        return """
+        MERGE INTO $quotedTableName AS Target
```

Comment:
The part that I don't like here is that the merge query is duplicated, same for column handling and escaping. I suspect this somehow doubles the maintenance effort.